operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,425 @@
+
+	// Get the first available CEB
+	for _, ceb := range c.cacheCeb {
+		// Note: If CEB's is marked for delete or no CEP's present in it don't add
+		// any more new CEP's in it.
+		if len(ceb.ceb.Endpoints) >= c.maxCepsInCeb || len(ceb.ceb.Endpoints) == 0 ||
+			c.cebsToSync.isCebMarkedForDelete(ceb.ceb.GetName()) {
+			continue
+		}
+		return ceb, false
+	}
+
+	// Allocate a new ceBatch and return
+	return c.createCeb(""), true
+}
+
+// InsertCepInCache is used to Insert CEP in local cache, this may result in Creating a New
+// CEB object or Updating an existing CEB object.
+func (c *cebManagerFcfs) InsertCepInCache(cep *cilium_v2.CoreCiliumEndpoint) (string, error) {
+
+	// Check the given cep is already packed in any of the CEB.
+	// if yes, Update a ceb with the given cep object.
+	if cebName, ok := c.cacheCepMap.get(getCepNameFromCCEP(cep)); ok {
+		queueCep(cep, c.cacheCeb[cebName])
+		c.cebsToSync.updateAggregator(cebName, CebUpdate)
+		return cebName, nil
+	}
+
+	cb, isNewCeb := c.findCeb(cep)
+
+	// Cache CEP name with newly allocated CEB.
+	c.cacheCepMap.insert(getCepNameFromCCEP(cep), cb.ceb.GetName())
+
+	// Queue the CEP in CEB
+	queueCep(cep, cb)
+
+	// If it isNewCeb, update Aggregator to create a ceb
+	if isNewCeb {
+		c.cebsToSync.updateAggregator(cb.ceb.GetName(), CebCreate)
+		return cb.ceb.GetName(), nil
+	}
+
+	c.cebsToSync.updateAggregator(cb.ceb.GetName(), CebUpdate)
+	return cb.ceb.GetName(), nil
+}
+
+// RemoveCepFromCache is used to remove CEP from local cache, this may result in
+// Updating an existing CEB object.
+func (c *cebManagerFcfs) RemoveCepFromCache(cep *cilium_v2.CoreCiliumEndpoint) error {
+
+	// Check in local cache, if a given cep is already batched in one of the ceb.
+	// and if exists, delete cep from ceb.
+	cepName := getCepNameFromCCEP(cep)
+	log.WithFields(logrus.Fields{
+		"cep-name": cepName,
+	}).Debug("Remove CEP from local cache")
+	if cebName, ok := c.cacheCepMap.get(cepName); ok {
+		ceb, err := c.cacheCeb[cebName]
+		if err {
+			if ceb == nil || ceb.ceb == nil {
+				log.Info("Valid cep Mapping, but no matching CEP in CEB")
+				return nil
+			}
+		}
+
+		for i, ep := range ceb.ceb.Endpoints {
+			if getCepNameFromCCEP(&ep) == getCepNameFromCCEP(cep) {
+				ceb.backendMutex.Lock()
+				// Insert deleted CoreCEP in removedCeps
+				ceb.removedCeps[getCepNameFromCCEP(&ep)] = struct{}{}
+				ceb.ceb.Endpoints =
+					append(ceb.ceb.Endpoints[:i],
+						ceb.ceb.Endpoints[i+1:]...)
+				ceb.backendMutex.Unlock()
+				break
+			}
+		}
+		log.WithFields(logrus.Fields{
+			"ceb-name":  cebName,
+			"cep-name":  cepName,
+			"cep-count": len(ceb.ceb.Endpoints),
+		}).Debug("Removed cep from ceb")
+
+		c.cacheCepMap.deleteCep(cepName)
+
+		if len(ceb.ceb.Endpoints) == 0 {
+			log.WithFields(logrus.Fields{
+				"ceb-name": cebName,
+			}).Debug("Remove CEB from local cache")
+			c.cebsToSync.updateAggregator(cebName, CebDelete)
```

Comment:
You said in another comment that a CEB will only be put in `DELETE` state if it is succesfully deleted in the apiserver, then what does this do? Doesn't this put the CEB in `DELETE` state?

Reply:
My bad, we should remove these lines. This isn't required.